### PR TITLE
Settings backend

### LINKED ISSUE #2296: Improve and extend `common-settings`
`common-settings` is for now a bit barren but we can easily see how it can be expanded to be a full-fledged configuration panel.

Now the problem is that it's very fragile and it has broken so many times in the past.  Indeed, since it's mostly made of arbitrary forms, there are barely any checks done, ever.

Besides, for now we just write to `auto-config.lisp` and we don't update the live instances.

I believe we could hit 2 birds with one stone: by updating the live instances, we might be able to perform more checking.

To sum up, each settings should be applied at 3 levels (optionally letting the user choose where):
- On the existing buffers.
- On the new buffers (`customize-instance` and `customize-hook`).
- For the next session (`auto-config`).

------

Some very useful settings:

- [ ] Ensure that user-defined (or extension-defined) keyschemes are listed.
- [ ] Dynamically preview theme styles. 
- [ ] Place "Edit user files" at the top.
- [ ] Restore session: always, ask, never.
- [ ] Make default browser, see #2439.
- [ ] Spell check language (it's a mode, should we have per-mode settings?)
- [ ] Download directory.
- [ ] Default modes.
- [ ] Proxy (also a mode)
- [ ] "set default new buffer URL" should have suggestions like "nyxt:new", "nyxt:dashboard".
- [ ] Customize search engines.
- [ ] Cookie settings.
- [ ] Intelligent Tracking Protection (it's part of `reduce-tracking-mode`, but maybe we should factor it out).
- [ ]  [Text editing keybindings](http://tiborsimko.org/emacs-bindings-in-gtk-apps.html).
- [ ] Font families 
- [ ] Removal of the data files (clear history, as one use-case for that).
- [ ] WebKit cache clearing _per category — cookies, history, cache, and a dozen others_.
- [ ] [WebKitSettings](https://webkitgtk.org/reference/webkit2gtk/stable/WebKitSettings.html), like buffer background and default font, lots of tasty stuff there. We may need to somehow propagate those to all buffers, maybe even maintaining a general set of [WwebKitSettings](https://webkitgtk.org/reference/webkit2gtk/stable/WebKitWebView.html#webkit-web-view-set-settings) to apply by default.

### PR DESCRIPTION
**Pull request by hgluka**
# Description

This is a continuation of (a part of) #2408.

Add a `setting` package for defining and applying configuration of Nyxt.
A setting can update a target class (by changing a slot value or by applying an arbitrary function to it) in the following places:
- the auto-config file
- a user-defined set of "live" instances
- all new instances (that will be opened in the current session)

This should provide a solid, convenient and testable back-end for configuring Nyxt.

Resolves #2296.

# Discussion

Right now, I'm working on properly finding a writer for a slot. There is the `closer-mop:slot-definition-writers` function, but that only works for *direct* slots, meaning that, for example, it doesn't work very well for `web-buffers`, since most of its slots are inherited.

cc @Ambrevar

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [ ] Git hygiene:
  - I have pulled from master before submitting this PR
  - There are no merge conflicts.
- [ ] I've added the new dependencies as:
  - ASDF dependencies,
  - Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - and Guix dependencies.
- [ ] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [ ] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [ ] Documentation:
  - All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - I have updated the existing documentation to match my changes.
  - I have commented my code in hard-to-understand areas.
  - I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
    - Changelog update should be a separate commit.
  - I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [ ] Compilation and tests:
  - My changes generate no new warnings.
  - I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - I ran the tests locally (`(asdf:test-system :nyxt)` and `(asdf:test-system :nyxt/gi-gtk)`) and they pass.

### CODE REVIEW

File: nyxt.asd [hunk 1]
```diff
@@ -220,6 +220,7 @@ The renderer is configured from NYXT_RENDERER or `*nyxt-renderer*'."))
                 :depends-on ("Core modes" "Modes")
                 :components
                 ((:file "help")
+                 (:file "setting")     ; REVIEW: Move to other module?
```
Comment (jmercouris) [inline nyxt.asd:223]: when you say module, are you using the Python notation to mean file?

**Comment by aartaka** [comment]
> That comment was left by Ambrevar, but I assume he was talking about whether the `setting` file fits into the "Help" module along with `help`, `manual`, `changelog`, etc.

Right! That's an ASDF module Pierre must've meant.

File: source/user-classes.lisp
Comment (hgluka) [inline source/user-classes.lisp:1]: The issue I was trying to solve here is that `auto-configure` uses `customize-instance`, 
while `extend-configuration` uses `customize-hook`. This means that when you apply a setting 
with `:auto-config-p t` and then restart Nyxt, applying another setting to that slot will have no effect on new instances.

Still, this seems to me like a very big change that could have unforeseen consequences.
What do you think? @Ambrevar

**Comment by aartaka** [comment]
> The issue I was trying to solve here is that `auto-configure` uses `customize-instance`, 
> while `extend-configuration` uses `customize-hook`. This means that when you apply a setting 
> with `:auto-config-p t` and then restart Nyxt, applying another setting to that slot will have no effect on new instances.

That's a perfect analysis of the issue. We should've moved to customize-hook (and/or define-configuration) a long time ago.

I'm pretty sure it'll make things simpler and more predictable, with only a couple of cornercases (that are present in define-configuration anyway) staying broken.

(These corner-cases are mainly the handler order and redefinition problems: if you redefine the buffer keyscheme to be Emacs, then VI, then back to Emacs—all in one file—then it might end up with either Emacs or VI. That's because customize-hook handler order is not guaranteed (but it should be pretty intuitive in most cases.) Anyway, this handler order problem happens regardless of customize-instance/customize-hook dichotomy, so let's ignoreit for now.)

**Comment by hgluka** [comment]
Hey, I've been working on this for about a month and I'm kind of getting stuck, so (as @jmercouris advised) I wanted to update everyone/ask for feedback and then maybe take a break from this PR and do something else before coming back to it.

The good part is that the `apply-setting` method works, it can update current instances, new instances and auto-configure. It can probably be used as is in the `common-settings` page and it can be tested just like `define-configuration` with it's own tests.

There are two big issues I haven't solved:
1. When we update current and future instances, we use objects (like `theme:+dark-theme+`, for example) directly, so those need to be serialized for `auto-configure`. I'm not sure if there's a good, general way to do this. There's also @aartaka's suggestion to rewrite it to use hooks instead of `customize-instance`, but that's mostly straightforward.
2. The settings work well enough for `browser` and `buffer` configuration but using them in the `describe-*` pages for arbitrary classes and slots is more difficult. It would work right now if we ignore live instances and just update new instances and auto-configure on those pages, but that's not an ideal solution.

If anyone has ideas for solving those two issues, let me know!

File: source/user-classes.lisp
Comment (aartaka) [reply source/user-classes.lisp:1]: >This means that when you apply a setting
with :auto-config-p t and then restart Nyxt, applying another setting to that slot will have no effect on new instances.
>Still, this seems to me like a very big change that could have unforeseen consequences.
What do you think? @Ambrevar

I'm late to the party, but the change would only make the behavior more predictable and compatible with other `define-configuration` settings. Especially after #3163.

It's more confusing when one configures Nyxt via Common Settings, then adds some setting via `define-configuration` in `config.lisp` only to see that it's not applied whatever value they pick. So using `customize-hook` is a bugfix, not a breaking change.

**Comment by aartaka** [comment]
> The good part is that the `apply-setting` method works, it can update current instances, new instances and auto-configure. It can probably be used as is in the `common-settings` page and it can be tested just like `define-configuration` with it's own tests.

That's a huge progress! 

> 1. When we update current and future instances, we use objects (like `theme:+dark-theme+`, for example) directly, so those need to be serialized for `auto-configure`. I'm not sure if there's a good, general way to do this. There's also aartaka's suggestion to rewrite it to use hooks instead of `customize-instance`, but that's mostly straightforward.

So rewriting it with hooks solves the problem? I'm not sure I get what the thing is about, can you expand on it?

> 2. The settings work well enough for `browser` and `buffer` configuration but using them in the `describe-*` pages for arbitrary classes and slots is more difficult. It would work right now if we ignore live instances and just update new instances and auto-configure on those pages, but that's not an ideal solution.

This problem is equivalent to "list all the instances of all classes" problem. Which, as far as I digged into the problem (I did, for quite some time) is hard and non-portable. SBCL has `sb-vm:map-allocated-objects` and that's as close to to properly searching instances as it can be. But other implementations don't have exported APIs for that. I'm pretty sure that CCL has something, but my search had no result yet. Allegro likely has some functionality for that, but it's proprietary and really obscure... Even though we are mostly building on SBCL, using its non-portable APIs is not a cool move.

We can hack up around this problem by keeping count on all the `user-class`-typed objects. We already hook into `make-instance` for these, so it shouldn't be hard. This solves the problem for all the Nyxt-related classes/instances at least. We can disable the configuration buttons for other classes—that'll be more honest anyway, because we actually can't guarantee any control over these.

This approach would be somewhat similar to what `*inspected-values*` do. Maybe it's a good opportunity to finally switch to implementation-provided IDs (#2991) and simply track all Nyxt objects by their IDs and their instances in some global table.

**Comment by hgluka** [comment]
> > 1. When we update current and future instances, we use objects (like `theme:+dark-theme+`, for example) directly, so those need to be serialized for `auto-configure`. I'm not sure if there's a good, general way to do this. There's also aartaka's suggestion to rewrite it to use hooks instead of `customize-instance`, but that's mostly straightforward.
> 
> So rewriting it with hooks solves the problem? I'm not sure I get what the thing is about, can you expand on it?

No, sorry for the confusion, those two things are only related because they have to do with `auto-configure`.
The issue I'm having is with serialization. When we try to configure a slot like `default-new-buffer-url`, the value is a string. We can set the slot value directly and we can write it to the `auto-config` file and get this:
```lisp
(defmethod customize-instance ((browser browser) &key)
  (setf (slot-value browser 'default-new-buffer-url) "https://example.com"))
```
That's good, everything works.
The problem comes up when we try to set the slot value to a more complex object, `theme` is a good example, but I assume there are others. 
This is the `auto-config` file when we try to configure the `theme` slot for the `browser` to `theme:+dark-theme+`:

```lisp
(defmethod customize-instance ((browser browser) &key)
  (setf (slot-value browser 'theme) #<theme:theme {1008530413}>))
```

We need the object there to be serialized, so that the file can be loaded on startup.

**Comment by aartaka** [comment]
I see now! A crude solution that I can come up with is comparing which slots in the `theme:theme` object (for the sake of example) had changed and construct a `make-instance` form with these. So, if the `background-color` and `on-background-color` swapped places, then the theme object might get serialized as:
``` lisp
(make-instance 'theme:theme :background-color "black" :on-background-color "white")
```

That's an incomplete solution, because current slot values for objects might be a result of some complex computation based on absolutely different initargs, but, I guess, that's the problem of the class with such complex computations :smile: 

I have a vague recollection of something like this instance diffing in our codebase, but that could well be an erroneous intuition.

**Comment by aartaka** [comment]
FYI: You can store raw Lisp objects [inside compiled files, and you can recover them while loading compiled files](https://www.reddit.com/r/Common_Lisp/comments/12dxdic/dumping_objects_into_compiled_files/). And it's [an ANSI standard feature](https://cl-community-spec.github.io/pages/Literal-Objects-in-Compiled-Files.html#Literal-Objects-in-Compiled-Files)!

**Comment by hgluka** [comment]
In case someone else picks this PR up, I'll do a write up of what I've figured out so far:

In talking with @aadcg last Thursday, he seems to think this is basically ready, but there are some issues left to fix and work to be done.

The biggest issue is with writing non-serializable slots to the `auto-config` file. This one can be solved by somehow modifying `auto-configure` to include serialization (like what @aartaka proposed in the above comments) or by circumventing the problem entirely. 

I'm not sure how easy or difficult the first option is, but the second one should be easier. 
In the `theme` example, it should be possible to pass the name of the object along with the value. The name can then be written to `auto-configure` and the value can be applied to live instances and added as a hook to future instances.
The only problem I see with that approach is generalizing it.

Outside of that, there is some junk to clean up, rebase on master (which has the new `common-settings` page), etc.
Because the `describe` pages and `common-settings` both use it, modifying `configure-slot` to use settings is the easiest way to incorporate these changes there.

**Comment by jmercouris** [comment]
There is too much wrong with this, and the system in general. I suggest /not/ merging this pull request. I need more time to think about how to do things properly. Possibly from the ground up.

**Pull request closed without merging**